source/math/FloatVectorList.ooc

```diff
@@ -381,4 +382,69 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> Int {
+		index := 0
+		thisPointer := (this pointer as Float*)
+		for (i in 1 .. this count)
+			if (thisPointer[index] < thisPointer[i])
+				index = i
+		index
+	}
+	correlation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart1 := 0.0f
+		lowerPart2 := 0.0f
+		thisMean := this mean
+		otherMean := other mean
+		thisPointer := (this pointer as Float*)
+		otherPointer := (other pointer as Float*)
+		for (i in 0 .. vectorLength) {
+			thisValue := thisPointer[i]
+			otherValue := otherPointer[i]
+			thisDiff := (thisValue - thisMean)
+			otherDiff := (otherValue - otherMean)
+			upperPart += thisDiff * otherDiff
+			lowerPart1 += thisDiff * thisDiff
+			lowerPart2 += otherDiff * otherDiff
+		}
+		upperPart / sqrt(lowerPart1 * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+		result := This new(this count)
+		thisPointer := (this pointer as Float*)
+		for (i in 0 .. this count) {
+			index := i + offset
+			value := (index >= 0 && index < this count) ? thisPointer[index] : filledValue
+			result add(value)
+		}
+		result
+	}
+	// offset < 0 means it shifts to left
+	calculateCrossCorrelation: func (other: This, offsetRange: Range, calculateOffsets := true) -> (This, This) {
+		crossCorrelation := This new(offsetRange count)
+		for (offset in offsetRange min .. offsetRange max + 1)
+			crossCorrelation add(this shift(offset) correlation(other, other count))
```

Comment:
I guess the compiler should be able to tell us when we create things and lose track of them in the same scope, but it would never be reliable enough. You'll have to keep using Valgrind, I'm afraid.